src/staker.ts

```diff
@@ -0,0 +1,195 @@
+import { MethodParameters, toHex } from './utils/calldata'
+import { defaultAbiCoder, Interface } from '@ethersproject/abi'
+import { abi } from '@uniswap/v3-staker/artifacts/contracts/UniswapV3Staker.sol/UniswapV3Staker.json'
+import { Pool } from './entities'
+
+export type FullWithdrawOptions = ClaimOptions & WithdrawOptions
+/**
+ * Represents a unique staking program.
+ */
+export interface IncentiveKey {
+  /**
+   * The token rewarded for participating in the staking program.
+   */
+  rewardToken: Token
+  /**
+   * The pool that the staked positions must provide in.
+   */
+  pool: Pool
+  /**
+   * The time when the incentive program begins.
+   */
+  startTime: BigintIsh
+  /**
+   * The time that the incentive program ends.
+   */
+  endTime: BigintIsh
+  /**
+   * The address which receives any remaining reward tokens at `endTime`.
+   */
+  refundee: string
+}
+
+/**
+ * Options to specify when claiming rewards.
+ */
+export interface ClaimOptions {
+  /**
+   * The id of the NFT
+   */
+  tokenId: BigintIsh
+
+  /**
+   * Address to send rewards to.
+   */
+  recipient: string
+
+  /**
+   * The amount of `rewardToken` to claim. 0 claims all.
+   */
+  amount: BigintIsh
+}
+
+/**
+ * Options to specify when withdrawing a position.
+ */
+export interface WithdrawOptions {
+  /**
+   * Set when withdrawing. The position will be sent to `owner` on withdraw.
+   */
+  owner: string
+
+  /**
+   * Set when withdrawing. `data` is passed to `safeTransferFrom` when transferring the position from contract back to owner.
+   */
+  data: string
+}
+
+export abstract class Staker {
+  public static INTERFACE: Interface = new Interface(abi)
+
+  protected constructor() {}
+  private static INCENTIVE_KEY_ABI =
+    'tuple(address rewardToken, address pool, uint256 startTime, uint256 endTime, address refundee)'
+
+  /**
+   *  To claim rewards, must unstake and then claim.
+   * @param incentiveKey The unique identifier of a staking program.
+   * @param options Options for producing the calldata to claim. Can't claim unless you unstake.
+   * @returns The calldatas for 'unstakeToken' and 'claimReward'.
+   */
+  private static encodeClaim(incentiveKey: IncentiveKey, options: ClaimOptions): string[] {
+    const calldatas: string[] = []
+    calldatas.push(
+      Staker.INTERFACE.encodeFunctionData('unstakeToken', [
+        this._encodeIncentiveKey(incentiveKey),
+        toHex(options.tokenId)
+      ])
+    )
+    const recipient: string = validateAndParseAddress(options.recipient)
+    calldatas.push(
+      Staker.INTERFACE.encodeFunctionData('claimReward', [
+        incentiveKey.rewardToken.address,
+        recipient,
+        toHex(options.amount)
+      ])
+    )
+    return calldatas
+  }
+
+  /**
+   * Only claims rewards. Must unstake, claim, and then restake.
+   */
+  public static collectRewards(incentiveKey: IncentiveKey, options: ClaimOptions): MethodParameters {
```

Comment:
let's also accept an array of `IncentiveKey`s here in addition to a singleton. this lets us support >=1 incentive for a single `tokenId` in all the public methods.

one strangeness that remains is technically you might want to specify an `amount` (or even `recipient`) per incentive, rather than globally, but i think we can live with that limitation for now in the expectation that most people won't want to do that.